# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/CoalescedTouchData.cpp
# Commit: 5718f7f3067e
# Full Hash: 5718f7f3067eff42e097a6013acd82fbe6aa299a
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2025-05-15 22:58:00
# Regressor Bug: 1550462
# File Overlap Count: 1
# Description:
#   Bug 1550462 - part 2: Make `PresShell::HandleEvent` dispatch preceding `pointerrawupdate` event r=smaug,dom-core,edgar
#   
#   This patch tries to dispatch `ePointerRawUpdate` with
#   `PresShell::EventHandler::DispatchPrecedingPointerEvent` as same as usual
#   pointer events.
# ==============================================================================

diff -r aae164cff6aa -r 5718f7f3067e dom/ipc/CoalescedTouchData.cpp
--- a/dom/ipc/CoalescedTouchData.cpp	Thu May 15 03:20:28 2025 +0000
+++ b/dom/ipc/CoalescedTouchData.cpp	Thu May 15 03:20:29 2025 +0000
@@ -28,6 +28,7 @@
     PointerEventHandler::InitPointerEventFromTouch(*event, aEvent, *touch);
     event->mFlags.mBubbles = false;
     event->mFlags.mCancelable = false;
+    event->convertToPointerRawUpdate = false;
   }
 }
 
@@ -47,10 +48,9 @@
     MOZ_ASSERT(mCoalescedInputEvent->mModifiers == aEvent.mModifiers);
     MOZ_ASSERT(mCoalescedInputEvent->mInputSource == aEvent.mInputSource);
 
-    for (size_t i = 0; i < aEvent.mTouches.Length(); i++) {
-      const RefPtr<Touch>& touch = aEvent.mTouches[i];
+    for (const RefPtr<Touch>& touch : aEvent.mTouches) {
       // Get the same touch in the original event
-      RefPtr<Touch> sameTouch = GetTouch(touch->Identifier());
+      const RefPtr<Touch> sameTouch = GetTouch(touch->Identifier());
       // The checks in CoalescedTouchData::CanCoalesce ensure it should never
       // be null.
       MOZ_ASSERT(sameTouch);
@@ -58,6 +58,7 @@
       MOZ_ASSERT(!sameTouch->mCoalescedWidgetEvents->mEvents.IsEmpty());
       if (!sameTouch->Equals(touch)) {
         sameTouch->SetSameAs(touch);
+        sameTouch->convertToPointerRawUpdate = touch->convertToPointerRawUpdate;
         WidgetPointerEvent* event =
             sameTouch->mCoalescedWidgetEvents->mEvents.AppendElement(
                 WidgetPointerEvent(aEvent.IsTrusted(), ePointerMove,
@@ -72,6 +73,18 @@
   }
 }
 
+void CoalescedTouchData::NotifyTouchRawUpdateOfHandled(
+    const WidgetTouchEvent& aEvent) {
+  if (IsEmpty()) {
+    return;
+  }
+  for (const RefPtr<Touch>& touch : aEvent.mTouches) {
+    if (const RefPtr<Touch> sameTouch = GetTouch(touch->Identifier())) {
+      sameTouch->convertToPointerRawUpdate = false;
+    }
+  }
+}
+
 bool CoalescedTouchData::CanCoalesce(const WidgetTouchEvent& aEvent,
                                      const ScrollableLayerGuid& aGuid,
                                      const uint64_t& aInputBlockId,